ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore/Util.hs

```diff
@@ -0,0 +1,97 @@
+import System.FilePath ((</>), takeDirectory, takeFileName)
+import System.FileLock (SharedExclusive (..), withFileLock)
+import System.IO (hClose, openBinaryTempFile)
+
+import qualified Data.ByteString.Lazy as LBS (hPut, readFile)
+
+import Ariadne.System.FileMode (ensureModeIs, mode600)
+import Ariadne.System.Other (whenPosix)
+import Ariadne.Wallet.Cardano.Kernel.Keystore.Types
+  (UserSecret(..), StorageDecodingError(..), usWallets, usFilePath)
+
+
+lockFilePath :: FilePath -> FilePath
+lockFilePath = (<> ".lock")
+
+-- | Creates empty temporary storage file for keyStore testing purposes
+getTempKeystorePath :: MonadIO m => m FilePath
+getTempKeystorePath = do
+    tempDir <- liftIO getTemporaryDirectory
+    let dir = tempDir </> "ariadne-test"
+    liftIO $ createDirectoryIfMissing False dir
+    liftIO $ createEmptyFile $ dir </> "keystore.json"
+
+createEmptyFile :: FilePath -> IO FilePath
+createEmptyFile fp = writeFile fp mempty $> fp
+
+-- | Creates empty keystore file and initialise user secret from default value
+initialiseUserSecret :: FilePath -> IO UserSecret
+initialiseUserSecret fp = do
+    _ <- createEmptyFile fp
+    whenPosix $ ensureModeIs mode600 fp
+    return $ def & usFilePath .~ fp
+
+-- | Reads userSecret from the given file.
+-- If the file does not exist or is empty return a default userSecret instance
+peekUserSecret :: FilePath -> IO UserSecret
+peekUserSecret fp = do
+    exists <- doesFileExist fp
+    if exists then
+        readUserSecret fp
+    else
+        initialiseUserSecret fp
+
+-- | Reads userSecret from file, assuming that file exists,and has mode 600.
+-- If file exists but is empty return the default userSecret instance.
+-- Throws exception in other case
+readUserSecret :: FilePath -> IO UserSecret
+readUserSecret fp = do
+    withReadLock fp $ do
+        content <- LBS.readFile fp
+        if null content then
+            pure $ def & usFilePath .~ fp
+        else do
+            either (throwM . StorageDecodingError . toText)
+                   (pure . set usFilePath fp)
+                   (eitherDecode content)
+
+-- | writes userSecret to file
+writeUserSecret :: UserSecret -> IO ()
+writeUserSecret us = do
+    let fp = us ^. usFilePath
+    withFileLock (lockFilePath fp) Exclusive $ const (writeRaw fp us)
+
+writeRaw :: FilePath -> UserSecret -> IO ()
+writeRaw path us = do
+    -- On POSIX platforms, openBinaryTempFile guarantees that the file
+    -- will be created with mode 600.
+    -- If openBinaryTempFile throws, we want to propagate this exception,
+    -- hence no handler.
+    (tempPath, tempHandle) <-
+        openBinaryTempFile (takeDirectory path) (takeFileName path)
+
+    LBS.hPut tempHandle (encode us) `finally` hClose tempHandle
+    hClose tempHandle
+    renameFile tempPath path
+
+-- | Whether stored UserSecret has empty keys
+isEmptyUserSecret :: UserSecret -> Bool
```

Comment:
This seems like the only function in this module that is not related to I/O. Furthermore, it is not used anywhere in this module. Perhaps it should be moved elsewhere?

Reply:
Agreed!